docs/design/provisioning.md

```diff
@@ -0,0 +1,143 @@
+				PersistentVolumeSelector:map[string]string{
+					"storage-class":"gold",
+				},
+			},
+		},
+	}
+	
+```
+
+#### Design Consideration
+
+`StorageClass` has been proposed as a field on PersistentVolumeClaim.  An equivalent of that field is currently stewing as an annotation in the current version of dynamic provisioning.
+
+`PersistentVolumeSelector` on PersistentVolumeClaim has been proposed as a means to select specific volumes, akin to a pod's NodeSelector.
+
+Having both fields is confusing.  Consider this on-premise installation of Kube:
+
+> Admin Joe configures "Gold" to dynamically provision Ceph volumes.  Joe also creates "Silver" volumes by manually provisioning a large number of NFS exports on expensive hardware (SSDs, highly available, stringent backup policy).  Lastly, Joe creates "Bronze" NFS exports on older hardware with slower discs.
+
+> How does a user request Gold/Silver/Bronze?   Using the StorageClass field for "Gold" while requiring the use of VolumeSelector for Silver/Bronze requires a user to know the difference between volumes.
+
+Using just VolumeSelector gives a user a single path to understand.  Admins retain the ability to manually create storage.  Labels on the volume (or the volume creator) are matched with the claim's volume selector.
+
+An additional benefit of using a selector for everything is the ability to deprecate PersistentVolumeAccessModes.  These fields on PV are descriptors only.  They have no functional value other than to describe the capabilities of the volume.  AccessModes on a PVC represent only a way to select certain volumes.  Admins can use labels to describe a volume's capabilities (e.g, ReadWriteMany) and users would select volumes accordingly.
+
+
+
+
+
+#### Other approaches considered:
+
+* Single ConfigMap with each key in Data representing 1 storage class, the value is opaque json.  Good for versioning the blob but it seemed unwieldy.
+* Many ConfigMaps with "storage-class" as a key in the Data.  Difficult to validate unique storage names.
+* Name of ConfigMap as name of storage class. Naming validation is enforced by API server.  Requires PVC.Spec.StorageClass (or similar) with 1:1 mapping of claim to provisioner.
+
+
+
+### Plugin Parameters
+
+The example above contains `"params": "${OPAQUE-JSON-MAP}"`, which is a string representing a JSON map of parameters.   Each provisionable volume plugin will document and publish the parameters it accepts.
+
+Plugin params can vary by whatever attributes are exposed by a storage provider.  An AWS EBS volume, for example, can be a general purpose SSD (gp2), provisioned IOPs (io1) for speed and throughput, and magnetic (standard) for low cost.  Additionally, EBS volumes can optionally be encrypted.  Configuring each volume type with optional encryption would require 6 distinct provisioners.  Administrators can mix and match any attributes exposed by the provisioning plugin to create distinct storage classes.
+
+
+## Controller behavior
```

Comment:
I think we need to consider the user experience for both the ADMIN and the DEVELOPER when defining the storage and consuming it. 
With using the selector rather than the class (where class is unique per the API) we yield to the possibility of duplicate of the storage description therefore making it difficult for the user DEVELOPER to determine the storage they are consuming. 
If we assume a mixed use environment where we have both statically created storage and dynamically provisioned storage, the path could provide the opposite of what the DEVELOPER is expecting based on the selector. Since the selector will just select on the first one, if there are duplicates, how will the user ever be able to claim against the second storage?

It also put a lot of responsibility on the admin to deeply understand the system and makes it hard for the scaling use case where there are 100-1000 of PVs created. Today we won't provide a good interface or tooling to assist the ADMIN in seeing the topology of the current pool and with the selector not using any validation, I just don't see how we can have consistent paths for each operation.